src/include/ceph_fs.h

```diff
@@ -790,6 +790,13 @@ int ceph_flags_to_mode(int flags);
 #define CEPH_CAP_LOCKS (CEPH_LOCK_IFILE | CEPH_LOCK_IAUTH | CEPH_LOCK_ILINK | \
 			CEPH_LOCK_IXATTR)
 
+/* cap masks async dir operations */
+#define CEPH_CAP_DIR_CREATE    CEPH_CAP_FILE_CACHE
+#define CEPH_CAP_DIR_UNLINK    CEPH_CAP_FILE_RD
```

Comment:
So if I'm given CEPH_CAP_DIR_UNLINK on a directory, what else do I need in order to unlink a dentry in that directory? I'll also need a valid dentry for the target, but the kernel client already does a lookup on it beforehand, so we can be reasonably sure that the dentry is present.

That seems like a workable model, and I Iike the basic lock cache concept.

What I'd really like to see is some sort of written spec for this that client authors can follow. It doesn't have to be anything formal, but I think we need to clearly document what each of the above caps is actually granting the client, and what else (if anything) it needs to acquire before performing these operations.

Also, are there ever situations where we have DIR_CREATE but not DIR_UNLINK? How much different is the mds locking for those cases?

Reply:
DIR_CREATE and DIR_UNLINK are separate from each other. Locking for create requires policy lock (for subtree file layout),  unlink does not need that lock.